types/basetypes/dynamic_type.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package basetypes
+
+import (
+	"context"
+	"errors"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// DynamicTypable extends attr.Type for dynamic types. Implement this interface to create a custom DynamicType type.
```

Comment:
Not sure where it would be appropriate to link this, but might be useful to have link(s) which provide further information on what a dynamic type is for, for instance, why you would want to use a dynamic type, and under what circumstances this would make sense, and be allowed